I want to participate in your project

sergeyblagodarnyy opened the pull request:
Hello, Mixaill. I appreciate your work and want to participate in the improvements. I added the following improvements. Pull them out please, if you agree, or say what you do not like.

 - Added Examples to build;
 - Changed default build test option to ON;
 - Added ${HEADERS} to library P7 sources, because QtCreator IDE sees it
in this case only;
 - Link pthread and rt libraries on linux build, because it needs it;

sergeyblagodarnyy committed:
- Added Examples to build;
 - Changed default build test option to ON;
 - Added ${HEADERS} to lobrary P7 sources, because QtCreator IDE sees it
in this case only;
 - Link pthread and rt libraries on linux build, because it needs it;

sergeyblagodarnyy committed:
- Added platform header files to build. Now it is showed in Qt Creator
IDE

sergeyblagodarnyy committed:
Added an ## to P7 macros for using in without additional arguments. Like
this:
P7_ERROR(my_module, "my error text");

Mixaill commented on Headers/P7_Trace.h:
```diff
@@ -79,73 +79,73 @@ enum eP7Trace_Level
                                                                  i_hModule,\
                                                                  i_pFormat,\
-                                                                 __VA_ARGS__)
+                                                                 ##__VA_ARGS__)
```
`##__VA_ARGS__` is a GCC extension which is, AFAIK, is not supported in MSVC.

In future, C++2a will solve this problem with something like `foo(bar, __VA_OPT__(,) __VA_ARGS__`)

sergeyblagodarnyy replied:
Solved 12e1dc0ad3a6c06be9e7115aebbc2950b650e6ff

Mixaill commented on Examples/Cpp/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+cmake_policy(SET CMP0063 NEW)
```
`CMP0063` was introduced in CMake 3.3. You need to change `cmake_minimum_required` in the root `CMakeLists.txt` file.

sergeyblagodarnyy replied:
Solved 67b5d75203d73c1bf0440c3e1ac9c1e622680200

sergeyblagodarnyy committed:
- Change non-standart macros ##__VA_ARGS__ to __VA_ARGS__
 - Delete an odd argument for using log macroses without arguments. Like this: P7_LOG_ERROR(p7_modules::socket, "A message");

sergeyblagodarnyy committed:
delete cmake_policy statements

Mixaill commented on Headers/P7_Trace.h:
```diff
@@ -72,14 +72,13 @@ enum eP7Trace_Level
+                                                            (tUINT16)__LINE__,\
+                                                            __FILE__,\
+                                                            __FUNCTION__,\
+                                                            __VA_ARGS__)
```
@sergeyblagodarnyy looks like `i_pFormat` is missing

sergeyblagodarnyy replied:
Yes, I did it intentionally, because, an expressions like:  LOG_ERROR(p7_modules::socket, "A message"); compiles with error. My compiler gcc expects more arguments after i_pFormat. It treats __VA_ARGS__ as "one or more arguments".

Now, I've modified the cpp test to demonstrate this. 098f2553b8da377c9cc755179583b34937888d43
This test doesn't compile on my side. The macro ##__VA_ARGS__  solves the problem, but you say it isn't crossplatform.

So, if you have any other idea how to solve the problem, I' ll be glad to see it.

Mixaill replied:
OK, got it

sergeyblagodarnyy committed:
- Added to cpp test some expression to demonstrate problem in log without additional arguments.
 - Deleted the i_pFormat arguments in macros do solve the problem
